sch_cake.c

```diff
@@ -1619,12 +1619,20 @@ static unsigned int cake_drop(struct Qdisc *sch, struct sk_buff **to_free)
 	return idx + (tin << 16);
 }
 
+static inline __be16 cake_skb_proto(const struct sk_buff *skb)
+{
+	if (skb->protocol == htons(ETH_P_8021Q))
+		return vlan_eth_hdr(skb)->h_vlan_encapsulated_proto;
+
+	return skb->protocol;
+}
+
 static u8 cake_handle_diffserv(struct sk_buff *skb, u16 wash)
 {
 	int wlen = skb_network_offset(skb);
 	u8 dscp;
 
-	switch (tc_skb_protocol(skb)) {
+	switch (cake_skb_proto(skb)) {
```

Comment:
Fix the other usage of tc_skb_protocol() as well?

Reply:
Do you mean change tc_skb_protocol() in upstream? Sounds good, but need to test all current usecases, and there are some difference in semantics already: in sch_cake we need to strip ether proto down to real protocols, other modules may rely on ETH_P_8021Q value, which will never be returned.

Reply:
I found that some code in upstream already relies on ETH_P_8021Q/ETH_P_8021AD (e. g. https://elixir.bootlin.com/linux/v5.7/source/net/sched/act_csum.c#L590 ), so the choises are to leave this as inline function (I prefer this, and it seems like other possible values should be handled here), or to extend handling of ETH_P_8021Q/e.t.c. in cake_handle_diffserv().

Reply:
No I just meant also change the other place we call that helper in sch_cake itself :)